### PR TITLE
[SMS-212] Added all storage constraints

### PR DESCRIPTION
**Pull request by abhishek0208**
### Description
Translated all storage constraints to linoopy

### Checklist
- [ ] Dependencies install correctly in a clean environment and code executes;
- [ ] Test coverage extended; created tests fail without the change (if possible);
- [ ] All tests passing;
- [ ] Commits follow a type convention (e.g. https://gist.github.com/brianclements/841ea7bffdb01346392c#type);
- [ ] Extended the README, documentation and/or docstrings, if necessary;

### CODE REVIEW

**Review by edwardxtg — APPROVED** [review]
Looking good.

The only thing I can think of adding is a solve test which includes storage, i'll add another ticket for this.

EDIT: Looks like storage is included in the utopia yaml example.

**Comment by edwardxtg** [comment]
@abhishek0208 @Lkruitwagen 
I've seen the tests are failing here as the test from yaml is passing an out of date yaml file, utopia.yaml.

This example has storage so would be a good example to keep as the test. I'll update it so that the test no longer gives infeasibility errors.

**Comment by Lkruitwagen** [comment]
yes - whatever is going wrong at the moment should `raise` as an error - we shouldn't be able to build this current Utopia model.

**Comment by edwardxtg** [comment]
@abhishek0208 @Lkruitwagen @Tomkourou 

Storage is now working and ready for review.

File: tz/osemosys/model/constraints/capacity_adequacy_a.py [hunk 1]
```diff
@@ -52,20 +56,20 @@ def add_capacity_adequacy_a_constraints(ds: xr.Dataset, m: Model) -> Model:
...
     )
-    con = m["AccumulatedNewCapacity"] - new_cap.where(mask).sum("BUILDYEAR") == 0
+    con = m["AccumulatedNewCapacity"] - lex["NewCapacity"].where(mask).sum("BUILDYEAR") == 0
     m.add_constraints(con, name="CAa1_TotalNewCapacity")
 
     con = m["TotalCapacityAnnual"] - m["AccumulatedNewCapacity"] == ds["ResidualCapacity"].fillna(0)
     m.add_constraints(con, name="CAa2_TotalAnnualCapacity")
 
-    RateOfTotalActivity = m["RateOfActivity"].sum(dims="MODE_OF_OPERATION")
+    # RateOfTotalActivity = m["RateOfActivity"].sum(dims="MODE_OF_OPERATION")
```
Comment (Lkruitwagen) [inline tz/osemosys/model/constraints/capacity_adequacy_a.py:69]: delete please!

**Review by Lkruitwagen — APPROVED** [review]
just one bit of housekeepin!

File: tz/osemosys/model/linear_expressions/reserve_margin.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
+from typing import Dict
+
+import xarray as xr
+from linopy import LinearExpression, Model
+
+
+def add_lex_reserve_margin(ds: xr.Dataset, m: Model, lex: Dict[str, LinearExpression]):
```
Comment (Tomkourou) [inline tz/osemosys/model/linear_expressions/reserve_margin.py:7]: Would be good to add docstrings here.

File: tz/osemosys/model/linear_expressions/storage.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
+from typing import Dict
+
+import xarray as xr
+from linopy import LinearExpression, Model
+
+
+def add_lex_storage(ds: xr.Dataset, m: Model, lex: Dict[str, LinearExpression]):
```
Comment (Tomkourou) [inline tz/osemosys/model/linear_expressions/storage.py:7]: same here.

**Review by Tomkourou — APPROVED** [review]

**Pull request merged**